### DISCUSSION TITLE
Test a function with a passed in struct that can be mocked

### DISCUSSION THREAD

**Discussion by Ben-PH** [asomers/mockall#478] [Questions] [answered]
We are introducing mocks into our codebase, and I'm running into a problem.

We are testing a function that takes in a struct as an argument, but we want to mock that struct. If we use the instructions (mockall_double, etc...), the compiler type-errors (Expected Foo, got MockFoo). 

The solution for now, is to mock `FooTrait`, and `impl Foo for FooTrait`, then make the argument generic:

```rust

fn uses_foo<F: Foo>(some_arg: (), foo: F) {
    ....
}
```

...this is okay-ish, but as we start using mockall throughout our codebase, we will start getting things like

```rust
struct LotsOfComponents<F: FooComponentTrait, B: BarComponentTrait, A: AlphaWorkerTrait, B: BravoWorkerTrait, ....> {
  foo: F,
  foo_worker: FW,
  some_name: A,
  another_name: B,
  running_out_of_names: C,
  how_many_generics_do_we_need: HMGDWN,
  ...
}
```
For reference: https://github.com/massalabs/massa/pull/3745#discussion_r1165197117

**Comment by Ben-PH** [comment]
I've created an issue that hypothesizes a workable approach> #479

**Comment by asomers** [comment]
You must be doing something wrong.  What you describe is exactly what `#[mockall_double]` is intended for.  I think you're probably  importing your struct in two different ways in different modules.  You're probably doing `use mymod::MyStruct` in one place, and `#[mockall_double] use mymod::MyStruct` somewhere else.  You must be consistent.

**Reply by Ben-PH** [reply]
I've stripped away test-conditional compilation, re-exports, etc, but I still get errors. I keep getting confusing errors.

Would it be possible to mock structs such that you still use the same type, only it redefines the struct such that it has a field that holds a mock-version of itself, and redefines the impls to use that mock-version?

**Reply by asomers** [reply]
> I've stripped away test-conditional compilation, re-exports, etc, but I still get errors. I keep getting confusing errors.
> 
> Would it be possible to mock structs such that you still use the same type, only it redefines the struct such that it has a field that holds a mock-version of itself, and redefines the impls to use that mock-version?

Absolutely not.  Rust does not work that way.

**Reply by Ben-PH** [reply]
so this:

```rust
#[internal_mock]
struct Foo {
   bar: u32
}

#[internal_mock]
impl Foo {
    fn uses_bar(&self) -> u32 {
        self.bar
    }
}
```
...couldn't expand to
```rust
struct Foo {
    mocked: MockFoo
}
impl Foo {
    fn uses_bar(&self) -> {
        self.mocked.uses_bar()
    }
```
???

**Reply by asomers** [reply]
Oh, it could expand to that all right.  But it wouldn't solve your problem, because this `Foo` would be a different type than the original `Foo`.  You would get the same compiler errors, but they would be more confusing because the two types would have the same name.  You would also get some "duplicate definition for struct Foo" errors, unless you put the new Foo into a submodule.  And if you did that, then they wouldn't have the same name anymore.

**Reply by Ben-PH** [reply]
how about this:

...but there is only ever one `Foo`. it's either the unexpanded one, or the expanded one.

**Reply by asomers** [reply]
That wouldn't solve your problem, and may well make it worse, if you arrange your code typically.  Because no code, even in the same module that defines Foo, would be able to access the real Foo in `test` mode.

**Reply by Ben-PH** [reply]
so what if it expands to:

```rust
struct Foo {
    bar: u32,
    mocked: Option<MockFoo>,
}
```
If it's to be used as a real `Foo`, then the mocked is set to None

**Comment by Ben-PH** [comment]
I'm revisiting our codebase.

usage:
```rust
    let mut mocked1 = NetworkCommandSender::default();
    let mut mocked2 = NetworkCommandSender::default();
    mocked2
        .expect_get_bootstrap_peers()
        .times(1)
        .returning(|| Ok(get_peers()));

    mocked1.expect_clone().return_once(move || mocked2);
```

imported with:
```rust
#[mockall_double::double]
use massa_network_exports::network_controller::NetworkCommandSender;
```
mocked in the `network_controller` mod:
```rust
#[cfg_attr(any(test, feature = "testing"), mockall::automock)]
impl NetworkCommandSender {
```

I tried re-exporting with mock_double in the structs parent module, then just importing naivly everywhere else (no use of mockdouble except at the re-export) and that seems to work :)

**Reply by Ben-PH** [reply]
...but this no longer works:

```rust
    let mut mocked1: NetworkCommandSender = NetworkCommandSender::default();
    let mut mocked2: NetworkCommandSender = NetworkCommandSender::default();
```

error:
```

error[E0599]: no function or associated item named `default` found for struct `NetworkCommandSender` in the current scope
   --> massa-bootstrap/src/tests/scenarios.rs:171:67
    |
171 |     let mut mocked1: NetworkCommandSender = NetworkCommandSender::default();
    |                                                                   ^^^^^^^ function or associated item not found in `NetworkCommandSender`

error[E0599]: no function or associated item named `default` found for struct `NetworkCommandSender` in the current scope
   --> massa-bootstrap/src/tests/scenarios.rs:172:67
    |
172 |     let mut mocked2: NetworkCommandSender = NetworkCommandSender::default();
```

I try something like this:
```rust
#[cfg(test)]
mockall::mock! {
    pub NetworkCommandsender {}
    impl Default for NetworkCommandSender {
        fn default() -> Self;
    }
}
#[cfg_attr(test, mockall::automock)]
impl NetworkCommandSender {
```

I get conflicting implementation error if i try to test-combile the buffer, and `default` is not implemented if I try to run a test that uses the `mockall_double` re-export.

If I do this:
```rust
#[cfg(test)]
mockall::mock! {

    pub NetworkCommandSender {
        async fn node_ban_by_ids(&self, ids: Vec<NodeId>) -> Result<(), NetworkError>;
        async fn node_ban_by_ips(&self, ips: Vec<IpAddr>) -> Result<(), NetworkError>;
        async fn add_to_whitelist(&self, ips: Vec<IpAddr>) -> Result<(), NetworkError>;
        async fn remove_from_whitelist(&self, ips: Vec<IpAddr>) -> Result<(), NetworkError>;
        async fn node_unban_by_ids(&self, ids: Vec<NodeId>) -> Result<(), NetworkError>;
        async fn node_unban_ips(&self, ips: Vec<IpAddr>) -> Result<(), NetworkError>;
        async fn send_block_info(&self, node: NodeId, info: Vec<(BlockId, BlockInfoReply)>) -> Result<(), NetworkError>;
        async fn ask_for_block_list(&self, list: HashMap<NodeId, Vec<(BlockId, AskForBlocksInfo)>>) -> Result<(), NetworkError>;
        async fn send_block_header(&self, node: NodeId, header: SecuredHeader) -> Result<(), NetworkError>;
        async fn get_peers(&self) -> Result<Peers, NetworkError>;
        async fn get_network_stats(&self) -> Result<NetworkStats, NetworkError>;
        async fn get_bootstrap_peers(&self) -> Result<BootstrapPeers, NetworkError>;
        async fn send_operations(&self, node: NodeId, operations: Vec<SecureShareOperation>) -> Result<(), NetworkError>;
        async fn announce_operations(&self, to_node: NodeId, batch: OperationPrefixIds) -> Result<(), NetworkError>;
        async fn send_ask_for_operations(&self, to_node: NodeId, wishlist: OperationPrefixIds) -> Result<(), NetworkError>;
        async fn send_endorsements(&self, node: NodeId, endorsements: Vec<SecureShareEndorsement>) -> Result<(), NetworkError>;
        async fn node_sign_message(&self, msg: Vec<u8>) -> Result<PubkeySig, NetworkError>;
        fn default() -> Self;
    }



}

#[cfg(not(test))]
impl NetworkCommandSender {
    async fn node_ban_by_ids(&self, ids: Vec<NodeId>) -> Result<(), NetworkError> {
        self.0
            .send(NetworkCommand::NodeBanByIds(ids))
            .await
            .map_err(|_| NetworkError::ChannelError("could not send BanId command".into()))?;
        Ok(())
    }
```
...that doesn't help

**Reply by Ben-PH** [reply]
...I tried this in the parent module (re-export the struct if not test, otherwise make that module public)

```rust
#[cfg(test)]
pub mod network_controller;
#[cfg(not(test))]
pub use network_controller::NetworkCommandSender;
#[cfg(not(test))]
mod network_controller;
```

and import it like so:
```rust
#[cfg(test)]
#[cfg_attr(test, mockall_double::double)]
use massa_network_exports::network_controller::NetworkCommandSender;
```

and I get this:
```
   Compiling massa_bootstrap v0.1.0 (/home/ben/rust/massa/massa-bootstrap)
error[E0432]: unresolved import `massa_network_exports::network_controller::MockNetworkCommandSender`
  --> massa-bootstrap/src/tests/scenarios.rs:39:5
   |
39 | use massa_network_exports::network_controller::NetworkCommandSender;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^--------------------
   |     |                                          |
   |     |                                          help: a similar name exists in the module: `NetworkCommandSender`
   |     no `MockNetworkCommandSender` in `network_controller`
```

**Comment by asomers** [comment]
> #[cfg_attr(any(test, feature = "testing"), mockall::automock)]

This is a problem .  `#[mockall_double]` does not know about the "testing" feature.  If you want to use that, then you'll have to do the doubling yourself.  Fortunately, it's easy:
```rust
#[cfg(any(test, feature = "testing"))]
use massa_network_exports::network_controller::MockNetworkCommandSender as NetworkCommandSender;
#[cfg(not(any(test, feature = "testing")))]
use massa_network_exports::network_controller::NetworkCommandSender;
```
> error[E0599]: no function or associated item named `default` found for struct `NetworkCommandSender` in the current scope

This means that you tried to call the `default` method on the real struct, not the mock struct.  Mockall always implements `Default`.  Probably the above error was the cause of this one.

**Reply by Ben-PH** [reply]
I'm still getting the same problems when I remove the feature attribute, and only check if it's a test combile or not

**Reply by Ben-PH** [reply]
```rust
#[cfg_attr(test, mockall::automock)]
impl NetworkCommandSender {
```

and the following in both in both `scenarios.rs` and `server.rs`

```rust
#[cfg(test)]
use massa_network_exports::network_controller::MockNetworkCommandSender as NetworkCommandSender;
#[cfg(not(test))]
use massa_network_exports::network_controller::NetworkCommandSender;
```
i get this error:

```
error[E0432]: unresolved import `massa_network_exports::network_controller::MockNetworkCommandSender`
  --> massa-bootstrap/src/server.rs:41:5
   |
41 | use massa_network_exports::network_controller::MockNetworkCommandSender as NetworkCommandSender;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^------------------------^^^^^^^^^^^^^^^^^^^^^^^^
   |     |                                          |
   |     |                                          help: a similar name exists in the module: `NetworkCommandSender`
   |     no `MockNetworkCommandSender` in `network_controller`

error[E0432]: unresolved import `massa_network_exports::network_controller::MockNetworkCommandSender`
  --> massa-bootstrap/src/tests/scenarios.rs:38:5
   |
38 | use massa_network_exports::network_controller::MockNetworkCommandSender as NetworkCommandSender;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^------------------------^^^^^^^^^^^^^^^^^^^^^^^^
   |     |                                          |
   |     |                                          help: a similar name exists in the module: `NetworkCommandSender`
   |     no `MockNetworkCommandSender` in `network_controller`
```

**Reply by asomers** [reply]
This isn't a problem in Mockall.  This is a problem with your `use` statements.  Maybe you've got a `pub use` in a module somewhere.  Maybe you're missing `pub` on `NetworkCommandSender`.  I don't know.  But you need to debug this the same way that you would debug any other unresolved import error.

**Comment by asomers** [comment]
By any chance, is `NetworkCommandSender` defined in a different crate than server.rs ?

**Reply by Ben-PH** [reply]
Correct.

**Reply by Ben-PH** [reply]
...it's part of a multi-crate workspace.

**Reply by asomers** [reply, marked as answer]
Then that's your problem.  Because when you build a crate in `test` mode, its dependencies are not build in `test` mode.  So what you want to do is probably to create a `testing` feature for the `massa_network_exports` crate and export mock objects when that feature is set, not just when `test` is set.  Then in your consuming create, ensure that you use the dependency's `testing` feature whenever you build in `test` mode.  Or, combine the two crates into one.

**Reply by Ben-PH** [reply]
That makes sense now, and explains why I was getting the errors, and why I was getting confused by them.

I'll put together a documentation PR. Hopefully it will reduce the chances of you having to go through this back-and-forth again.

**Reply by zephod77** [reply]
Looks like I have exactly the same issue. Can you expand on your answer a bit more. I didn't understand how to export mock objects.
In my case I have a crate called common_messaging that has a lib.rs that exports 3 modules one of which is called RawMessage.
RawMessage has an associated function called read_from_skt defined like this:
```pub fn read_from_skt(zmq_skt: &MockableZmqSocket) -> Result<RawMessage, zmq::Error>```
MockableZmqSocket is defined like this:
```                                                                                     
#[cfg(not(test))]                                                                   
  use zmq::Socket as MockableZmqSocket;                                               
  #[cfg(test)]                                                                        
  use tests::MockSocket as MockableZmqSocket; 
```
and tests::MockSocket is defined:
```
mod tests {
...
mock! {
    pub Socket {...}
}
```
so how do I export MockableZmqSocket or just MockSocket?

**Reply by asomers** [reply]
Are you trying to define mock objects in one crate and use them in another?  Then you can't use `#[cfg(test)]`, because external crates are _never_ built in test mode.

**Reply by zephod77** [reply]
Correct. I have another crate that imports my RawMessage crate and it builds and runs fine but I need to unit test it.

I have a function that calls RawMessage::read_from_socket() and I get this error message when I run cargo test on it:
```
  --> src/net_agent.rs:84:60
   |
84 |             let rcvd_msg = match RawMessage::read_from_skt(&rep_skt) {
   |                                  ------------------------- ^^^^^^^^ expected struct `Socket`, found struct `MockSocket`
   |                                  |
   |                                  arguments to this function are incorrect
   |
   = note: expected reference `&Socket`
              found reference `&net_agent_unit_tests::MockSocket`
note: associated function defined here
  --> /devroot/acs_src/KACommonSrc/ltv_ka/rust/common_message/src/raw_message.rs:38:12
   |
38 |     pub fn read_from_skt(zmq_skt: &MockableZmqSocket) ->
```
which is the same issue as described in this discussion, correct?
The way I read you answer, is that I have to export MockSocket from my common_messaging crate using features. Is that correct? So I was asking about how to do that.

**Reply by asomers** [reply]
So src/net_agent.rs has imported MockSocket, but it's trying to use externally-defined RawMessage?  And since external crates are built without `#[cfg(test)]`, RawMessage wants a real Socket.  

The bottom line is that Rust is statically typed, and Mockall doesn't really do anything special with regards to typing.  You need to adjust your `use` statements to be consistent.  If you want common_messaging's consumers to be able to use either real Sockets or MockSockets, then you need to either export two versions of RawMessage, or else change read_from_skt's signature to take a trait object.

Or, change your project organization.  IMHO the most effective way to use mock objects is to organize your code in layers.  Use mock objects to unit test one layer with mock versions of the next lower layer's objects.  But don't let mock objects (or anything else, really) cross layers.  If you stick to that philosophy, then common_messaging won't need to export MockSocket.  Instead, it may export MockRawMessage, and net_agent.rs won't ever access Socket or MockSocket directly.

**Reply by zephod77** [reply]
Organizing my code into layers is what I am trying to do. RawMessage is the bottom layer and it has its own set of unit tests. Its read_from_skt() associated function takes a Socket reference which is defined as &MockableZmqSocket. 
MockSocket in RawMessage is only defined when running the RawMessage unit tests.

net_agent.rs imports RawMessage and it creates the socket that gets passed to a function that calls RawMessage::read_from_skt() so:
```
fn init() {
    ...
    let skt: MockableZmqSocket = ...
    spawn(move || foo(skt));
}
fn foo (skt: MockableZmqSocket) {
    loop {
        ...
        rm = RawMessage.read_from_skt(&skt);
        ...
    }
}
```
The net_agent unit tests have their own MockSocket so that init() can be tested and I have mocked RawMessage:
```
mock! {
    RawMessage {
        pub fn read_from_skt(  ??? ) -> Result<RawMessage, zmq::Error>;
    }
}
```
If I make the ??? = MockSocket, that doesn't work because read_from_skt() is expecting a MockableZmqSocket as a zmq::Socket and if I make it = zmq::Socket that doesn't work because foo() takes a MockableZmqSocket which is a MockSocket when compiled as test.  As you say, read_from_skt() need a real Socket when testing net_agent. foo()'s skt parameter needs to be a zmq::Socket when testing foo() but a MockSocket when testing init().

I think you are suggesting that RawMessage export MockRawMessage but I don't really understand that as MockRawMessage is only defined in net_agent when testing.

Thanks for taking the time to look at this.

**Reply by asomers** [reply]
I said that when I thought that you were using RawMessage from the common_messaging crate .  Using mock! within net_agent is actually better.  Do you really have three "Socket" types?  Socket, MockSocket, and MockableZmqSocket?  And "MockableZmqSocket" is only used in non-test configurations?  What is "mockable" about MockableZmqSocket?

**Reply by zephod77** [reply]
MockableZmqSocket is the name I use in the conditional compilation
```
#[cfg(not(test))]
use zmq::Socket as MockableZmqSocket;
#[cfg(test)]
use tests::MockSocket as MockableZmqSocket;

fn bar() {
    let s: MockableZmqSocket = ...
}
```

**Reply by asomers** [reply]
I'm sorry, but I really can't find your problem without seeing your entire program, both crates.  I'm sure that this isn't any bug in Mockall.  You just need to be consistent about which types you use where.  But in general:
* Avoid passing mock objects between crates
* Absolutely don't do anything in `#[cfg(test)]` and expect that other crates will see it.
* Try to layer your code cleanly.  Ideally, never pass a mock object through more than one layer.

**Reply by zephod77** [reply]
I was preparing a minimum version of my code to send to you when I realized that I was mocking the wrong thing. Once I mocked my RawMessage instead of the socket to test my foo() function, I had a lot more success ;)
I am now having an issue with synchronizing the tests that mock an associated function but that is another subject.

**Reply by asomers** [reply]
Good news.  I'm glad you resolved your problem.  As for synchronization, associated functions are a global resource.  So you need to protect access to them just like any other global resource.  See https://docs.rs/mockall/latest/mockall/#static-methods
